File: scala/src/test/scala/ad/Chapter2Test.scala
package ad

import org.scalatest._

import com.ad.Chapter2._

import scala.Function

class Chapter2Test extends FlatSpec with Matchers {

  // Exercise 2.1
  "fib" should "produce expected value for small inputs" in {
    fib(1) should be (0)
    fib(2) should be (1)
    fib(3) should be (1)
    fib(4) should be (2)
    fib(5) should be (3)
    fib(6) should be (5)
  }

  "fib" should "produce expected value for large inputs" in {
    fib(30) should be (514229)

  }


  "fib" should "should fail for n below 1" in {
    intercept[IllegalArgumentException] {
      fib(0)
    }

    intercept[IllegalArgumentException] {
      fib(-1)
    }
  }

  // Exercise 2.2
  "isSorted" should "be true for empty string array" in {
    isSorted[String](Array(), (a, b) => a.compareTo(b) > 0 ) shouldBe true
  }

  "isSorted" should "be true for empty int array" in {
    isSorted[String](Array(), (a, b) => a.compareTo(b) > 0 ) shouldBe  true
  }

  "isSorted" should "be true for array of strings sorted in ascending order" in {
    isSorted[String](Array("1", "2", "3"), (a, b) => b.compareTo(a) > 0 ) shouldBe true
  }

  "isSorted" should "be true for array of strings sorted in descending order" in {
    isSorted[String](Array("3", "2", "1"), (a, b) => a.compareTo(b) > 0 ) shouldBe true
  }

  "isSorted" should "be false for array of strings not in any sorted order" in {
    isSorted[String](Array("3", "5", "1"), (a, b) => a.compareTo(b) > 0 ) shouldBe false
  }

  "isSorted" should "be true for array of ints sorted in ascending order" in {
    isSorted[Int](Array(1, 2, 3), (a, b) => a < b ) shouldBe true
  }

  "isSorted" should "be true for array of ints sorted in descending order" in {
    isSorted[Int](Array(4, 3, 2), (a, b) => a > b ) shouldBe true
  }

  "isSorted" should "be false for array of int not in any sorted order" in {
    isSorted[Int](Array(4, 3, 6, 9), (a, b) => a > b ) shouldBe false
  }
 //  // Exercise 2.3
 // def curry[A, B, C](f: (A, B) => C) : A => (B => C)
 "curry" should " a function with two arguments return a function that takes one argument and returns a function " in {
   val f :(Int, Int) => Int = (x, y) => x + y
   val cf = curry(f)
   cf(2)(3) shouldBe 5
 }

  "curry" should " a function with two arguments return a function that can be partially applied " in {
    val f :(Int, Int) => Int = (x, y) => x + y
    val pcf = curry(f)(2)
    pcf(3) shouldBe 5
  }


  // Exercise 2.4
 // def uncurry[A, B, C](f: A => B => C )
  "uncurry" should " a convert a curried function (f: A => B => C ) into two arguments function f:(A, B) => C " in {
    val f :(Int, Int) => Int = (x, y) => x + y
    val cf = curry(f)
    cf(2)(3) shouldBe 5
    uncurry(cf)(2, 3) shouldBe 5
  }

 // exercise 2.5 . Compose
  "compose" should " satisfy composition f o g i.e g(x) -> y , f(y) -> z then f(g(x)) -> z" in {
    val g : String => String = s => s.reverse
    val f : String => Int = s => s.toInt

    g("123") shouldBe "321"
    f("123") shouldBe 123

    compose(f, g)("123") shouldBe 321
    compose(f, g)("123") shouldBe f(g("123"))
  }




}

File: haskell/src/Chapter2Test.hs
import Test.Hspec
import Test.QuickCheck
import Data.List
import Chapter2 ( fib,
                  isSorted,
                  curry,
                  uncurry,
                  compose
                     )
import Prelude hiding(curry, uncurry, compose)
main = hspec $
 describe "Chapter2" $ do
  describe "fib" $ do
   it "works for first 10 numbers" $ do
    map (fib) [1..10] `shouldBe` [Just 0, Just 1, Just 1, Just 2, Just 3, Just 5, Just 8, Just 13, Just 21, Just 34]
   it "returns Nothing for negative numbers" $ do
    fib (-1) `shouldBe` Nothing


  describe "isSorted" $ do
   it "works for sorted lists of int in ascending order" $ do
    isSorted [1..10] (<) `shouldBe` True
   it "works for sorted lists of int in descending order" $ do
    isSorted [10,9..1] (>) `shouldBe` True
   it "works for unsorted lists of int in no order" $ do
    isSorted (2:[10,9..1]) (>) `shouldBe` False
   it "works for numbers increasing and decreasing order " $ do
    isSorted [1, 2, 1, 2] (<) `shouldBe` False
   it "works for sorted lists of strings in ascending order" $ do
    isSorted (map (show) [1..9]) (<) `shouldBe` True
   it "works for sorted lists of strings in descending order" $ do
    isSorted (map (show) [9,8..1]) (>) `shouldBe` True
   it "works for unsorted lists of strings" $ do
    isSorted ["A", "c", "d"] (>) `shouldBe` False

  describe "curry" $ do
   it "converts function f(A, B) -> C in f(A) -> (B -> C) " $ do
   (Chapter2.curry f) 2 3 `shouldBe` 5


  describe "uncurry" $ do
   it "converts function f(A -> B -> C) in f((A, B) -> C) " $ do
   (Chapter2.uncurry g)(2, 3) `shouldBe` 5

  describe "compose" $ do
     it "converts function f(A -> B) and g(B -> C) in h(A -> C) " $ do
     (Chapter2.compose reverse sort)[2, 5, 3]  `shouldBe` [5, 3, 2]

   where f (a, b) = a + b
         g  a b = a + b

